packages/flutter_tools/lib/src/ios/devices.dart

```diff
@@ -451,22 +424,7 @@ class IOSDevice extends Device {
       }
 
       _logger.printTrace('Application launched on the device. Waiting for observatory port.');
-      final FallbackDiscovery fallbackDiscovery = FallbackDiscovery(
-        logger: _logger,
-        portForwarder: portForwarder,
-        protocolDiscovery: observatoryDiscovery,
-        flutterUsage: globals.flutterUsage,
-        pollingDelay: fallbackPollingDelay,
-        vmServiceConnectUri: _vmServiceConnectUri,
-      );
-      final Uri localUri = await fallbackDiscovery.discover(
-        assumedDevicePort: assumedObservatoryPort,
-        device: this,
-        usesIpv6: ipv6,
-        hostVmservicePort: debuggingOptions.hostVmServicePort,
-        packageId: packageId,
-        packageName: FlutterProject.current().manifest.appName,
-      );
+      final Uri localUri = await observatoryDiscovery.uri;
```

Comment:
This will hang forever when the app crashes because the log reader will never find the VM service URI log line.  That will be fixed on iOS 13+ with https://github.com/flutter/flutter/pull/68844 on top of this change.  However we need to figure out how to handle <=iOS 12 crashes on launch that are syslog scanning.

I will work a bit on unifying the code paths.  I tried the debugging dance on iOS 9 and it didn't work, but I didn't investigate too hard because the syslog scanning was working.

Reply:
stable v 2.5.3, run on iphone x and 8 with iOS 15+ , after install and launching still got observatory timeout log and then stuck for about 90s to get the writing analzed variants log, then main page shows, here's the log blow  
Running Xcode build...
Xcode build done.                                           21.0s
Installing and launching...
iOS Observatory not discovered after 30 seconds. This is taking much longer than expected...
(lldb) 2021-11-18 10:40:33.805817+0800 Runner[5156:2537590] Writing analzed variants.
Debug service listening on ws://127.0.0.1:49248/s6cQ7AOhIgE=/ws